### PR TITLE
Native IPFS support in Brave behind a feature flag

### PR DESCRIPTION
**Pull request by bbondy**
Part of https://github.com/brave/brave-browser/issues/10220

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/ipfs/content_browser_client_helper.cc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/render_view_host.h"
+#include "extensions/browser/extension_registry.h"
+#include "extensions/common/constants.h"
+#include "extensions/common/extension_set.h"
+#include "net/base/escape.h"
+#include "url/gurl.h"
```
Comment (yrliou) [inline browser/ipfs/content_browser_client_helper.cc:31]: I believe we only need below headers.
```
#include "brave/browser/ipfs/content_browser_client_helper.h"

#include <string>
#include <utility>

#include "base/task/post_task.h"
#include "brave/common/url_constants.h"
#include "brave/common/pref_names.h"
#include "brave/components/ipfs/common/ipfs_constants.h"
#include "chrome/browser/external_protocol/external_protocol_handler.h"
#include "components/prefs/pref_service.h"
#include "components/user_prefs/user_prefs.h"
#include "content/public/browser/browser_context.h"
#include "content/public/browser/browser_task_traits.h"
#include "content/public/browser/browser_thread.h"
#include "content/public/browser/render_process_host.h"
#include "content/public/browser/render_view_host.h"
#include "url/gurl.h"
```

File: browser/infobars/ipfs_infobar_delegate.cc [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#include "brave/browser/infobars/ipfs_infobar_delegate.h"
+
+#include <memory>
+#include <utility>
+
+#include "brave/browser/brave_browser_process_impl.h"
+#include "brave/browser/ui/brave_pages.h"
+#include "brave/components/ipfs/browser/brave_ipfs_client_updater.h"
+#include "brave/components/ipfs/common/ipfs_constants.h"
+#include "brave/common/brave_wallet_constants.h"
```
Comment (yrliou) [inline browser/infobars/ipfs_infobar_delegate.cc:15]: remove this one.

File: browser/infobars/ipfs_infobar_delegate.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_INFOBARS_IPFS_INFOBAR_DELEGATE_H_
+#define BRAVE_BROWSER_INFOBARS_IPFS_INFOBAR_DELEGATE_H_
+
+#include "base/compiler_specific.h"
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/infobars/ipfs_infobar_delegate.h]: Removed in https://github.com/brave/brave-core/pull/5820/commits/45c6e70b2d94cc0968e9bf8aa331822b6fe647e5

File: browser/extensions/BUILD.gn [hunk 4]
```diff
@@ -111,6 +113,16 @@ source_set("extensions") {
     ]    
   }
 
+  if (ipfs_enabled) {
+    sources += [
+      "api/ipfs_api.cc",
+      "api/ipfs_api.h",
+    ]
+    deps += [
+      "//brave/components/ipfs/browser"
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/extensions/BUILD.gn]: Added in https://github.com/brave/brave-core/pull/5820/commits/45c6e70b2d94cc0968e9bf8aa331822b6fe647e5

File: browser/extensions/api/ipfs_api.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#include "base/json/json_writer.h"
+#include "base/values.h"
+#include "brave/browser/ipfs/ipfs_service.h"
+#include "brave/browser/ipfs/ipfs_service_factory.h"
+#include "brave/browser/profiles/profile_util.h"
+#include "brave/common/extensions/extension_constants.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/ipfs/browser/brave_ipfs_client_updater.h"
+#include "brave/components/ipfs/browser/features.h"
+#include "brave/components/ipfs/common/ipfs_constants.h"
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/extensions/api/ipfs_api.cc:16]: Added in 45c6e70

File: browser/infobars/ipfs_infobar_delegate.cc [hunk 6]
```diff
@@ -0,0 +1,95 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/infobars/ipfs_infobar_delegate.cc]: Resolved in https://github.com/brave/brave-core/pull/5820/commits/71aa380319fd4b4202e542b37d120195ffdfd8a9

File: browser/infobars/BUILD.gn [hunk 7]
```diff
@@ -40,6 +41,17 @@ source_set("infobars") {
     ]
   }
 
+  if (ipfs_enabled) {
+    sources += [
+      "ipfs_infobar_delegate.cc",
+      "ipfs_infobar_delegate.h",
+    ]
+    deps += [
+      "//brave/components/ipfs/common",
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/infobars/BUILD.gn:51]: Added in 45c6e70b2d94cc0968e9bf8aa331822b6fe647e5

File: browser/ipfs/content_browser_client_helper.cc (hunk 1 above)
Comment (yrliou) [reply browser/ipfs/content_browser_client_helper.cc:23]: Removed in 45c6e70b2d94cc0968e9bf8aa331822b6fe647e5

File: browser/infobars/ipfs_infobar_delegate.cc (hunk 2 above)
Comment (yrliou) [reply browser/infobars/ipfs_infobar_delegate.cc]: Removed in 45c6e70

File: browser/ipfs/content_browser_client_helper.cc [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+namespace {
+
+bool IsIPFSDisabled(content::BrowserContext* browser_context) {
+  auto* prefs = Profile::FromBrowserContext(browser_context)->GetPrefs();
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/ipfs/content_browser_client_helper.cc]: Sorry I meant to left this unchanged before hearing your feedback but I accidentally changed it already in 45c6e70, let me know if you want me to change this back.

File: browser/extensions/api/ipfs_api.cc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+#include "brave/browser/extensions/api/ipfs_api.h"
+
+#include <memory>
+#include <string>
+
+#include "base/json/json_writer.h"
+#include "base/values.h"
+#include "brave/browser/ipfs/ipfs_service.h"
+#include "brave/browser/ipfs/ipfs_service_factory.h"
+#include "brave/browser/profiles/profile_util.h"
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/extensions/api/ipfs_api.cc]: Removed in 45c6e70

File: browser/extensions/api/ipfs_api.cc [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+#include "content/public/browser/web_contents.h"
+#include "extensions/browser/extension_util.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace {
+
+ipfs::IpfsService* GetIPFSService(
+    content::BrowserContext* context) {
+  return ipfs::IpfsServiceFactory::GetInstance()
+      ->GetForContext(Profile::FromBrowserContext(context));
```
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/extensions/api/ipfs_api.cc]: Fixed in 45c6e70

File: browser/ipfs/content_browser_client_helper.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (bbondy) [reply browser/ipfs/content_browser_client_helper.cc]: I'm good w/ the `user_prefs` way

File: browser/autocomplete/brave_autocomplete_scheme_classifier.cc [hunk 11]
```diff
@@ -37,6 +38,10 @@ BraveAutocompleteSchemeClassifier::GetInputTypeForScheme(
       base::LowerCaseEqualsASCII(scheme, kBraveUIScheme)) {
     return metrics::OmniboxInputType::URL;
   }
+  if (base::IsStringASCII(scheme) &&
+      base::LowerCaseEqualsASCII(scheme, kBraveUIScheme)) {
+    return metrics::OmniboxInputType::URL;
+  }
```
Comment (bbondy) [inline browser/autocomplete/brave_autocomplete_scheme_classifier.cc:44]: Looks like an unintentional duplicated block, I'll clean this up.

**Comment by bbondy** [comment]
Approved reviewing Jocelyn's commits 
Can't mark it because I'm the PR author.

File: browser/ipfs/ipfs_service.cc [hunk 12]
```diff
@@ -189,8 +190,14 @@ std::unique_ptr<network::SimpleURLLoader> IpfsService::CreateURLLoader(
...
   request->method = "POST";
-  request->headers.SetHeader(net::HttpRequestHeaders::kOrigin,
-                             server_endpoint_.spec());
+
+  // Remove trailing "/".
+  std::string origin = server_endpoint_.spec();
+  if (base::EndsWith(origin, "/", base::CompareCase::INSENSITIVE_ASCII)) {
+    origin.pop_back();
+  }
+  request->headers.SetHeader(net::HttpRequestHeaders::kOrigin, origin);
```
Comment (bbondy) [inline browser/ipfs/ipfs_service.cc:199]: ++

**Pull request merged**